healsparse/io_map_fits.py

```diff
@@ -511,3 +537,62 @@ def _write_map_fits(hsp_map, filename, clobber=False, nocompress=False):
     else:
         # All other maps are not compressed.
         _write_filename(filename, c_hdr, s_hdr, hsp_map._cov_map[:], hsp_map._sparse_map)
+
+
+def _write_moc_fits(hsp_map, filename, clobber=False):
+    """
+    Write the valid pixels of a HealSparseMap to a multi-order component (MOC)
+    file.  Note that values of the pixels are not persisted in MOC format.
+
+    Parameters
+    ----------
+    hsp_map : `healsparse.HealSparseMap`
+        HealSparseMap with valid_pixels to output.
+    filename : `str`
+        Name of file to save
+    clobber : `bool`, optional
+        Clobber existing file?  Default is False.
+    """
+    import astropy.io.fits as fits
+
+    max_order = int(np.round(np.log2(hsp_map.nside_sparse)))
+    min_uniq_order = int(np.round(np.log2(hsp_map.nside_coverage)))
+
+    pixels = hsp_map.valid_pixels
+
+    uniq = 4*(4**max_order) + pixels
+
+    uniq_map = hsp_map.make_empty(hsp_map.nside_coverage, hsp_map.nside_sparse, dtype=np.float32)
+    uniq_map[pixels] = 1.0
+
+    # Loop over orders, degrade each time, and look for pixels with full coverage.
+    for uniq_order in range(max_order - 1, min_uniq_order - 1, -1):
+        uniq_map = uniq_map.degrade(2**uniq_order, reduction='sum')
+        pix_shift = np.right_shift(pixels, 2*(max_order - uniq_order))
+        # Check if any of the pixels at uniq_order have full coverage.
+        covered, = np.isclose(uniq_map[pix_shift], 4**(max_order - uniq_order)).nonzero()
+        if covered.size == 0:
+            # No pixels at uniq_order are fully covered, we're done.
+            break
+        # Replace the UNIQ pixels that are fully covered
+        uniq[covered] = 4*(4**uniq_order) + pix_shift[covered]
+
+    # Remove duplicate pixels
+    uniq = np.unique(uniq)
+
+    # Output to fits
+    tbl = np.zeros(uniq.size, dtype=[('UNIQ', 'i8')])
+    tbl['UNIQ'][:] = uniq
+
+    order = np.round(np.log2(tbl['UNIQ']//4)).astype(np.int32)//2
+    moc_order = np.max(order)
+
+    hdu = fits.BinTableHDU(tbl)
+    hdu.header['PIXTYPE'] = 'HEALPIX'
+    hdu.header['ORDERING'] = 'NUNIQ'
+    hdu.header['COORDSYS'] = 'C'
+    hdu.header['MOCORDER'] = moc_order
+    hdu.header['MOCTOOL'] = 'healsparse'
+    hdu.header['MOCVERS'] = '1.1'
```

Comment:
I guess this choice is due to `pymoc` using v1.1? Do you know when/if they plan to transition to 2.0?

Reply:
The 2.0 spec was just released (in March), and I have had a hell of a time figuring out the format because the documentation is unclear and the mocpy fits reading is buggy and at the moment I just didn't have the patience.  I wanted to get something in here right now that works, and expand from there if it proves useful.

Reply:
thanks. It makes sense. I agree to sticking to 1.1 for now, and see how things evolve. I've been looking for their FITS-IO routines, but they don't seem to be easy to find (I found something in rust that I guess they pull now from the MOC rust core package: https://github.com/cds-astro/mocpy/blob/95fe96b371504449cd4590c9d80049db66be4532/src/moc/src/deser/fits/mod.rs)

Reply:
Yup, it's down in rust ... and at that point I gave up looking further.  The present problem is that astropy will output (perfectly legal) FITS with `TFORM1  = 'K       '`, but the mocpy fits reader (at least the one used in the non-deprecated branch) demands `TFORM1  = '1K      '`.